ktor-server/ktor-server-netty/jvm/src/io/ktor/server/netty/cio/RequestBodyHandler.kt

```diff
@@ -26,36 +30,48 @@ internal class RequestBodyHandler(
 
     private val job = launch(context.executor().asCoroutineDispatcher(), start = CoroutineStart.LAZY) {
         var current: ByteWriteChannel? = null
+        var expectedLength = -1L
+        var written = 0L
         var upgraded = false
 
+        fun checkCurrentLengthAndClose() {
+            if (expectedLength == -1L || written == expectedLength) {
+                current?.close()
+                return
+            }
+
+            val message = "Unexpected length of the request body. Expected $expectedLength but was $written"
+            current?.close(IOException(message))
+        }
+
         try {
             while (true) {
-                @OptIn(ExperimentalCoroutinesApi::class)
                 val event = queue.tryReceive().getOrNull()
                     ?: run { current?.flush(); queue.receiveCatching().getOrNull() }
                     ?: break
 
                 when (event) {
                     is ByteBufHolder -> {
-                        val channel = current
-                            ?: throw IllegalStateException("No current channel but received a byte buf")
-                        processContent(channel, event)
+                        val channel = current ?: error("No current channel but received a byte buf")
+                        written += processContent(channel, event)
 
                         if (!upgraded && event is LastHttpContent) {
-                            current.close()
+                            checkCurrentLengthAndClose()
```

Comment:
Isn't it redundant, since this check is in `finally` block?

Reply:
Nope: the next message can be a new channel in the pipeline